### ISSUE TITLE
prompt/history pollution by `for` and `if` statements

### ISSUE THREAD

**Issue by Konstruktionist** [fish-shell/fish-shell#5501]
`fish --version`
  fish, version 3.0.0
  `echo $version`
  3.0.0
`uname -a`
Darwin kafka 18.2.0 Darwin Kernel Version 18.2.0: Mon Nov 12 20:24:46 PST 2018; root:xnu-4903.231.4~2/RELEASE_X86_64 x86_64

 `sh -c 'env HOME=$(mktemp -d) fish'`
This does change things. Fish behaves as expected.

---

When I do the following (either manually or pasted):

    for var in (set -n | grep "fish.*_color.*")
        echo $var (set_color $$var) $$var (set_color normal)
    end

and then arrow up to go through `history`, once I'm on this and start deleting everything gets deleted except the `f` from the `for` statement. I can then type `ls` `Enter`(prompt shows `fls`) and ls gets executed.

If I arrow up past the `for` into history all of the history  gets prepended by an f. So I get `fecho`, `fvim`, `fgit` etc.

Same happens with `if` blocks, then the `i` pollutes the history browsing.
It can be fixed by deleting the `for` or `if` from history, but that's kinda defeats the purpose of `history`

I've never noticed this before v3.0, so I think it's new.  
I can see this behaviour in iTerm2 v3.2.6 and Terminal v2.9.1

I've made an [asciinema](https://asciinema.org/a/220411) recording. On my machine it shows this happening, but on the uploaded version not.

I've also made a [mp4 recording](https://streamable.com/sonus) that does show what is happening.

Since after `sh -c 'env HOME=$(mktemp -d) fish'` doesn't show the problem I guess there is something wrong in my setup. `fish_prompt` most likely. But I have no clue what might be the problem here.  

[My github repository for fish stuff](https://github.com/Konstruktionist/fish)

My fish_prompt

````
function fish_prompt -d 'Write out the prompt'

  # What the prompt has to show
  #
  #   1. display username & host when in a ssh session
  #   2. username if we're logged in as root on localhost
  #   3. current directory
  #      a. shortened to 'x' chars per directory if path length is > 60 chars long
  #           where 'x' is defined by fish_prompt_pwd_dir_length (default = 1)
  #           This can be changed in config.fish
  #      b. otherwise show full path
  #   4. git info:  (from prompt_git_status) 
  #      a. git branch name (if we are in a git repo)
  #      b. wether or not the working copy is dirty
  #      c. git sha hash
  #      d. display up/down arrow if there is stuff to be pushed/pulled
  #   5. display prompt symbol (in red if last command exited with an error)

  set -l last_status $status
  set -l user (whoami)

  # Get the current working directory & simplify $HOME to ~
  set cur_wd (echo $PWD | sed -e "s|^$HOME|~|" -e 's|^/private||')
  # Do we have to shorten the string?
  if test (string length $cur_wd) -gt 60
    set -q fish_prompt_pwd_dir_length
    or set -l fish_prompt_pwd_dir_length 1
    set cur_wd (prompt_pwd)
  else
    set cur_wd $cur_wd
  end

  # 1. ssh session
  if test -n "$SSH_CONNECTION"
    set -l host (hostname -s)

    # color of user@host is set by $user, either red (root) or blue (non-root)
    if test "$user" = "root"
      set user (set_color -r red)"$user"
    else
      set user (set_color -r blue)"$user"
    end

    set user_host "$user@$host"
    echo -n $user_host; set_color normal ; echo -n ' '
  end

  # 2. Are we logged in as root on localhost?
  if test -z "$SSH_CONNECTION"
    if test "$user" = "root"
      set user (set_color -r red)"$user"
      echo -n $user; set_color normal; echo -n ' '
    else
      set user ''
      echo -n $user
    end
  end

  # 3. current directory
  set_color $fish_color_cwd; echo -n $cur_wd; set_color normal

  # 4. Git info
  prompt_git_status
  echo    # effectivly sets a newline (\n)

  # 5. Prompt symbol
  if test "$last_status" -ne 0
    set_color $fish_color_error # Red
  end
  echo -n '▶︎ '
  set_color normal

end
````

**Comment by zanchey** [comment]
Does removing the '▶︎' symbol from your prompt fix the problem?

If so, your terminal emulator and fish disagree on how wide that character is (see #4539). You might need to turn on "ambiguous characters are double-width" in your iTerm settings and set `fish_emoji_wish` to 2 - at least that fixes things for me in 3.0.0.

**Comment by Konstruktionist** [comment]
Yep, that fixed it. Thanks
